### PR TITLE
add ProvisionalProposal struct

### LINKED ISSUE #73: `PayjoinProposal` typestate is leaky
`prepare_psbt` requires no changes be done to the internal psbt after `apply_fee` is called. These two should be separated into something like `ProvisionalProposal`, from which inputs/outputs can be changed until `apply_fee` is called, which returns `PayjoinProposal` from which the Payjoin Proposal PSBT can be extracted.

### PR DESCRIPTION
**Pull request by jbesraa**
resolves #73

### CODE REVIEW

**Comment by DanGould** [comment]
🏋️‍♂️🤙

It might make more sense to pass a closure as a parameter than a trait we'd have to define and implement downstream. This is how check methods are done earlier in the module. Take a look at those and let me know which parts you need more clarity on

File: payjoin/Cargo.toml [hunk 1]
```diff
@@ -22,10 +22,10 @@ bip21 = "0.3.1"
 log = { version = "0.4.14"}
 rand = { version = "0.8.4", optional = true }
 url = "2.2.2"
+bitcoind = { version = "0.31.1", features = ["0_21_2"] }
```
Comment (DanGould) [inline payjoin/Cargo.toml:25]: PDK can't force downstream dependencies to use bitcoind, it has been generic enough to bring your own wallet and should continue to be.

File: payjoin/src/receive/mod.rs [hunk 2]
```diff
@@ -791,7 +808,34 @@ impl PayjoinProposal {
...
+        Ok(PayjoinProposal {
+            payjoin_psbt: self.payjoin_psbt,
+            owned_vouts: self.owned_vouts,
+            params: self.params,
+        })
+    }
+
+    pub fn finalize_proposal(
+        mut self,
+        wallet_process_psbt: impl Fn(&str) -> Result<WalletProcessPsbtResult, bitcoincore_rpc::Error>,
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:820]: This function signature has to be generic enough for downstream libraries without bitcoind to use it. It should be strongly typed. I imagine a signature like `impl Fn(&Psbt) -> Result<Psbt, Error>` would make the most sense. Perhaps the parameter should be owned.

File: payjoin/src/receive/mod.rs [hunk 3]
```diff
@@ -566,6 +567,20 @@ impl PayjoinProposal {
         self.params.disable_output_substitution
     }
 
+    pub fn get_owned_vouts(&self) -> Vec<usize> { self.owned_vouts.clone() }
+
+    pub fn get_payjoin_psbt(&self) -> Psbt { self.payjoin_psbt.clone() }
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:572]: is there a good reason to clone here instead of letting `owned_vouts() -> self.owned_vouts` and the same for payjoin_psbt?

File: payjoin/src/receive/mod.rs [hunk 4]
```diff
@@ -791,7 +803,37 @@ impl PayjoinProposal {
...
+        };
+        let payjoin_proposal_psbt = match Psbt::from_str(&payjoin_proposal_psbt.to_string()) {
+            Ok(p) => p,
+            Err(e) => return Err(RequestError::from(InternalRequestError::PsbtParseError(e))),
+        };
+
+        let payjoin_proposal = match self.prepare_psbt(payjoin_proposal_psbt) {
+            Ok(p) => p,
+            Err(e) => return Err(RequestError::from(e)),
+        };
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:835]: `.map_error` or even a new `to_request_error` might turn this into something less verbose

e.g. something like

```rs
let payjoin_proposal = self.prepare_psbt(payjoin_proposal_psbt).map_err(RequestError::from)?

/// or maybe it's .map_err(|e| e.into)? i'm not sure
```

File: payjoin-cli/src/app.rs [hunk 5]
```diff
@@ -308,23 +309,25 @@ impl App {
...
+                            None,
+                            None,
+                            Some(false),
+                        )
+                        .map(|res| res.psbt)
+                        .map_err(|e| Error::Server(e.into()))
+                },
+                Some(1),
+            )
+            .unwrap();
```
Comment (DanGould) [inline payjoin-cli/src/app.rs:329]: this error needs to be handled

File: payjoin-cli/src/app.rs [hunk 6]
```diff
@@ -286,20 +286,21 @@ impl App {
...
-            if let Ok(address) = bitcoin::Address::from_script(output_script, network) {
-                self.bitcoind
-                    .get_address_info(&address)
-                    .map(|info| info.is_mine.unwrap_or(false))
-                    .map_err(|e| Error::Server(e.into()))
-            } else {
-                Ok(false)
-            }
-        })?;
+        let mut provisional_payjoin: ProvisionalProposal =
```
Comment (DanGould) [inline payjoin-cli/src/app.rs:289]: is this explicit type required? Seems redundant since you're using such care with naming.

a nit would be to make check4's result `proposal` or `provisional_payjoin` instead of becoming a payjoin and turning into a provisional_payjoin

**Review by DanGould — COMMENTED** [review]
This is starting to look beautiful. The attention to detail and effort is paying off

File: payjoin-cli/src/app.rs (hunk 6 above)
Comment (jbesraa) [reply payjoin-cli/src/app.rs]: not really, removed.

created a new issue for check4 change here https://github.com/payjoin/rust-payjoin/issues/96, will handle that in a new pr as i dont wanna add more changes to this pr

**Comment by jbesraa** [comment]
@DanGould thanks so much for the input. I made the required changes, feel free to reiview (:

File: payjoin/src/receive/error.rs [hunk 7]
```diff
@@ -41,7 +41,9 @@ pub struct RequestError(InternalRequestError);
 #[derive(Debug)]
 pub(crate) enum InternalRequestError {
     Psbt(bitcoin::psbt::Error),
+    PsbtParseError(bitcoin::psbt::PsbtParseError),
     Base64(bitcoin::base64::DecodeError),
+    PsbtFinalizingError(&'static str),
```
Comment (DanGould) [inline payjoin/src/receive/error.rs:46]: This always returns the same kind of string, so it doesn't need to take a value at all

`InternalRequestError::PsbtFinalize` would be fine

File: payjoin/src/receive/mod.rs [hunk 8]
```diff
@@ -791,7 +803,32 @@ impl PayjoinProposal {
...
+    ) -> Result<PayjoinProposal, RequestError> {
+        let payjoin_proposal_psbt: &Psbt = self.apply_fee(min_feerate_sat_per_vb)?;
+        let payjoin_proposal_psbt = match wallet_process_psbt(payjoin_proposal_psbt) {
+            Ok(p) => p,
+            Err(_) => {
+                return Err(RequestError::from(InternalRequestError::PsbtFinalizingError(
+                    "wallet_process_psbt failed",
+                )));
+            }
+        };
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:826]:
```suggestion
        let psbt = wallet_process_psbt(payjoin_proposal_psbt)
            .map_err(|_| InternalRequestError::PsbtFinalizingError.into())?;
```

I think it's fine to call it plain `psbt` in this context, we know it's a payjoin_proposal method function.

File: payjoin/tests/integration.rs [hunk 9]
```diff
@@ -200,20 +200,24 @@ mod integration {
...
+                        .wallet_process_psbt(
+                            &bitcoin::base64::encode(psbt.serialize()),
+                            None,
+                            None,
+                            Some(false),
+                        )
+                        .map(|res| res.psbt)
+                        .map_err(|e| Error::Server(e.into()))
+                },
+                Some(1),
```
Comment (DanGould) [inline payjoin/tests/integration.rs:216]: magic numbers usually mean something has gone wrong somewhere. I think we should accept a FeeRate here instead of u64

File: payjoin/tests/integration.rs [hunk 10]
```diff
@@ -200,20 +200,24 @@ mod integration {
...
+            )
+            .unwrap();
+        let psbt = payjoin_proposal.get_payjoin_psbt();
```
Comment (DanGould) [inline payjoin/tests/integration.rs:219]: Is there a good reason to return a PayjoinProposal to hang onto instead of just the psbt straight away? there might be.

File: payjoin/src/receive/mod.rs [hunk 11]
```diff
@@ -550,8 +551,8 @@ impl OutputsUnknown {
 }
 
 /// A mutable checked proposal that the receiver may contribute inputs to to make a payjoin.
+
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:554]: stray line

File: payjoin/Cargo.toml [hunk 12]
```diff
@@ -28,4 +28,4 @@ env_logger = "0.9.0"
 bitcoind = { version = "0.31.1", features = ["0_21_2"] }
 
 [package.metadata.docs.rs]
-features = ["send", "receive"]
```
Comment (DanGould) [inline payjoin/Cargo.toml:31]: stray change here

**Review by DanGould — COMMENTED** [review]
Overall this is looking like a much less leaky way to handle typestate. I'm impressed.

I left a few more questions about the interface design

File: payjoin-cli/src/app.rs (hunk 6 above)
Comment (DanGould) [reply payjoin-cli/src/app.rs]: I'm a bit confused why you chose to make this a new issue, it seems this PR introduces the issue, no?

**Comment by jbesraa** [comment]
hey @DanGould thanks for the review. 
I made some small changes and wanna allocate a bigger time window for this to get to a better shape. I am trying to run the integration tests locally without success, any tips about that?
I tried with  ` cargo test --verbose --all-features --lib` and ` cargo test tests/integration.rs --verbose --all-features --lib` without success.

**Comment by DanGould** [comment]
`RUST_LOG=debug cargo test --package payjoin --test integration --features send --features receive --  integration::integration_test --exact --nocapture` works for me

File: payjoin/src/receive/error.rs (hunk 7 above)
Comment (jbesraa) [reply payjoin/src/receive/error.rs]: fixed!

File: payjoin-cli/src/app.rs [hunk 13]
```diff
@@ -318,23 +319,23 @@ impl App {
...
+        let payjoi_proposal = provisional_payjoin.finalize_proposal(
+            |psbt: &Psbt| {
+                self.bitcoind
+                    .wallet_process_psbt(
+                        &bitcoin::base64::encode(psbt.serialize()),
+                        None,
+                        None,
+                        Some(false),
+                    )
+                    .map(|res| Psbt::from_str(&res.psbt).unwrap())
```
Comment (jbesraa) [inline payjoin-cli/src/app.rs:333]: @DanGould been trying to tackle this `unwrap` without success. how would you approach it?

File: payjoin/tests/integration.rs (hunk 10 above)
Comment (jbesraa) [reply payjoin/tests/integration.rs:223]: yea we might need to return other stuff in the future so In that case it wouldnt be a breaking change

File: payjoin/Cargo.toml (hunk 12 above)
Comment (jbesraa) [reply payjoin/Cargo.toml:31]: so weird, cant find this change locally

**Comment by jbesraa** [comment]
hey @DanGould I made few changes and resolved #96 
we need to add a bit of docs to reflect this change, any other place than the code that I need to take into account?

File: payjoin-cli/src/app.rs (hunk 13 above)
Comment (DanGould) [reply payjoin-cli/src/app.rs:333]: This article gives a few different options for returning errors from iterators: https://doc.rust-lang.org/rust-by-example/error/iter_result.html

I think the main issue is that it it doesn't make sense for `finalize_proposal` to return a `RequestError`, because the request has already been checked by this point. The error is more of a receiver::Error::Server and could be returned inside that Box.

File: payjoin/tests/integration.rs (hunk 10 above)
Comment (DanGould) [reply payjoin/tests/integration.rs:223]: That makes sense. It might make more sense to name this `psbt()` since we know it's in the PayjoinProposal and `get_` is associated with returning from an index in rust by convention https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

File: payjoin/src/receive/mod.rs [hunk 14]
```diff
@@ -791,7 +807,25 @@ impl PayjoinProposal {
...
+        })
+    }
+
+    pub fn finalize_proposal(
+        mut self,
+        wallet_process_psbt: impl Fn(&Psbt) -> Result<Psbt, Error>,
+        min_feerate_sat_per_vb: Option<FeeRate>,
+    ) -> Result<PayjoinProposal, RequestError> {
+        let psbt = self.apply_fee(min_feerate_sat_per_vb)?;
+        let psbt = wallet_process_psbt(psbt).map_err(|_| {
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:823]: better to handle the specific error instead of discarding it by returning `Error` from finalize_proposal instead of `RequestError`. Then we don't have to convert after `prepare_psbt` either

File: payjoin-cli/src/app.rs [hunk 15]
```diff
@@ -291,22 +291,23 @@ impl App {
...
+                if let Ok(address) = bitcoin::Address::from_script(output_script, network) {
+                    self.bitcoind
+                        .get_address_info(&address)
+                        .map(|info| info.is_mine.unwrap_or(false))
+                        .map_err(|e| Error::Server(e.into()))
+                } else {
+                    Ok(false)
+                }
+            },
+        )?;
```
Comment (DanGould) [inline payjoin-cli/src/app.rs:308]: identify_receiver_outputs does not seem directly related to the check, so I think we lose some semantics by merging them into one step. Either way it's unrelated to the ProvisionalProposal struct and would be better suited to its own PR

File: payjoin/src/receive/mod.rs [hunk 16]
```diff
@@ -491,7 +491,8 @@ impl MaybeInputsSeen {
     pub fn check_no_inputs_seen_before(
         self,
         is_known: impl Fn(&OutPoint) -> Result<bool, Error>,
-    ) -> Result<OutputsUnknown, Error> {
+        is_receiver_output: impl Fn(&Script) -> Result<bool, Error>,
+    ) -> Result<ProvisionalProposal, Error> {
```
Comment (DanGould) [inline payjoin/src/receive/mod.rs:495]: I can see why these might be merged, but perhaps the checks should return `RequestError` (since if they fail, the request is bad) while the closure returns Error, since that's a problem with the server implementation. Either way this is separate from the ProvisionalProposal change

File: payjoin-cli/src/app.rs (hunk 13 above)
Comment (jbesraa) [reply payjoin-cli/src/app.rs]: ah changing that to Error made things so easier! thanks

File: payjoin-cli/src/app.rs (hunk 15 above)
Comment (jbesraa) [reply payjoin-cli/src/app.rs:308]: yea I see what u mean
maybe we can change the parent function name

Comment (DanGould) [reply payjoin-cli/src/app.rs:308]: I don't think it should be renamed rather not combined. It's unrelated to this PR.

**Comment by DanGould** [comment]
I've re-ordered the commits so that each one passes tests and changed the apply_fee param to `min_fee_rate` since it's using the `FeeRate` type and doesn't need to specify sats_per_byte.

@jbesraa you've made this receiver interface 10x cleaner at least ⭐️
